backends/cxxrtl/runtime/cxxrtl/cxxrtl.h

```diff
@@ -891,8 +892,21 @@ struct fmt_part {
 	}
 };
 
+// An object that can be passed to a `eval()` method in order to act on side effects.
+struct performer {
+	// Called to evaluate a Verilog `$time` expression.
+	virtual int64_t time() const { return 0; }
+
+	// Called to evaluate a Verilog `$realtime` expression.
+	virtual double realtime() const { return time(); }
+
+	// Called when a `$print` cell is triggered.
+	virtual void on_print(const std::string &output) { std::cout << output; }
+};
+
 // An object that can be passed to a `commit()` method in order to produce a replay log of every state change in
-// the simulation.
+// the simulation. Unlike `performer`, `observer` does not use virtual calls as their overhead is unacceptable, and
+// a comparatively heavyweight template-based solution is justified.
```

Comment:
Heavyweight as in compile-time heavyweight? Complexity-wise the templating doesn't seem that heavy.

Reply:
It requires you to have the `commit` method in the header, for one, so that it's visible to any potential call sites which need an observer. That is quite invasive compared to what was there before, which is a single virtual call (which we still have and which has a wrapper delegating to this template.)

There continues to be nothing particularly compile-time heavyweight in CXXRTL. Parsing STL includes takes about as much time as expanding a CXXRTL netlist for a RISC-V CPU.

Reply:
I wonder if there can be a heavily-instrumented design where the overhead of the virtual call per every enabled `$check` starts to matter, in any case it shouldn't matter for a typical design.

Reply:
This seems like something the indirect call predictor should cover well. For `commit()` though I have benchmarked it and the current implementation is just about the best I can do while still using pointers as identifiers (assigning identifiers persistently in the module would give additional 15% or so, which is desirable, but it's also kind of hard to do).